conn.go

```diff
@@ -604,7 +647,239 @@ func (c *Conn) handleData(arg string) {
+				err = ew
+				break
+			}
+			if nr != nw {
+				err = io.ErrShortWrite
+				break
+			}
+		}
+		if er != nil {
+			if er != io.EOF {
+				err = er
+			}
+			break
+		}
+	}
+
+	if consumed < n && err == nil {
+		// src stopped early; must have been EOF.
+		err = io.EOF
+	}
+
+	return consumed, err
+}
+
+func (c *Conn) handleBdat(arg string) {
+	args := strings.Split(arg, " ")
+	if len(args) == 0 {
+		c.WriteResponse(501, EnhancedCode{5, 5, 4}, "Missing chunk size argument")
+		return
+	}
+
+	if !c.fromReceived || len(c.recipients) == 0 {
+		c.WriteResponse(502, EnhancedCode{5, 5, 1}, "Missing RCPT TO command.")
+		return
+	}
+
+	last := false
+	if len(args) == 2 {
+		if !strings.EqualFold(args[1], "LAST") {
+			c.WriteResponse(501, EnhancedCode{5, 5, 4}, "Unknown BDAT argument")
+			return
+		}
+		last = true
+	}
+
+	// ParseUint instead of Atoi so we will not accept negative values.
+	size, err := strconv.ParseUint(args[0], 10, 32)
+	if err != nil {
+		c.WriteResponse(501, EnhancedCode{5, 5, 4}, "Malformed size argument")
+		return
+	}
+
+	if c.bdatStatus == nil && c.server.LMTP {
+		c.bdatStatus = c.createStatusCollector()
+	}
+
+	if c.bdatPipe == nil {
+		var r *io.PipeReader
+		r, c.bdatPipe = io.Pipe()
+
+		c.dataResult = make(chan error)
+
+		go func() {
+			defer func() {
+				if err := recover(); err != nil {
+					c.handlePanic(err, c.bdatStatus)
+
+					select {
+					case <-c.dataResult:
+						c.dataResult <- errPanic
+					default:
+					}
+
+					r.CloseWithError(errPanic)
+				}
+			}()
+
+			var err error
+			if !c.server.LMTP {
+				err = c.Session().Data(r)
+			} else {
+				lmtpSession, ok := c.Session().(LMTPSession)
+				if !ok {
+					err = c.Session().Data(r)
+					for _, rcpt := range c.recipients {
+						c.bdatStatus.SetStatus(rcpt, err)
+					}
+				} else {
+					err = lmtpSession.LMTPData(r, c.bdatStatus)
+				}
+			}
+
+			select {
+			case <-c.dataResult:
+				// handleBdat goroutine wants us to sent an error here.
+				c.dataResult <- err
+			default:
+			}
+
+			if err != nil {
+				r.CloseWithError(err)
+			} else {
+				r.Close()
+			}
+		}()
+	}
+
+	// Using copyN that returns amount of bytes read from c.text.R instead of
+	// amount of bytes written is important here because read may succeed and
+	// write not in case of early error returned by Session.Data. In this case
+	// io.CopyN call below needs to calculate correct amount of bytes to
+	// discard.
+	consumed, err := copyN(c.bdatPipe, c.text.R, int64(size))
+	if err != nil {
+		// Backend might return an error without consuming the whole chunk.
+		io.CopyN(ioutil.Discard, c.text.R, int64(size)-consumed)
+
+		code, enhancedCode, msg := toSMTPStatus(err)
+		c.WriteResponse(code, enhancedCode, msg)
+
+		if err == errPanic {
+			c.Close()
+		}
+
+		return
+	}
+
+	c.bytesReceived += int(consumed)
+	if c.server.MaxMessageBytes != 0 && c.bytesReceived > c.server.MaxMessageBytes {
+		c.WriteResponse(552, EnhancedCode{5, 3, 4}, "Max message size exceeded")
```

Comment:
Should we close the `bdatPipe` at this point?

Maybe we can do this before actually copying the data?

Reply:
Right, that check should be moved before io.Copy.